File: Leetcode-Python/UniqueBinarySearchTrees2.py
"""
Given n, generate all structurally unique BST's (binary search trees) that store values 1...n.

For example,
Given n = 3, your program should return all 5 unique BST's shown below.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
"""
from TreeNode import TreeNode


class Solution:
    # @param {integer} n
    # @return {TreeNode[]}
    def generateTrees(self, n):

        def __generateTrees(left, right):
            res = []
            if left > right:
                return [None]
            else:
                for mid in xrange(left, right + 1):
                    for left_child in __generateTrees(left, mid - 1):
                        for right_child in __generateTrees(mid + 1, right):
                            root = TreeNode(mid)
                            root.left = left_child
                            root.right = right_child
                            res.append(root)
                return res

        return __generateTrees(1, n)

if __name__ == "__main__":
    roots = Solution().generateTrees(3)

File: Java-By-Tag/src/main/UniqueBinarySearchTrees2.java
package src.main;

import apple.laf.JRSUIUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by yhf on 8/23/15.
 */

/**
 * Given n, generate all structurally unique BST's (binary search trees) that store values 1...n.
 *
 * For example,
 * Given n = 3, your program should return all 5 unique BST's shown below.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
 */

public class UniqueBinarySearchTrees2 {
    public List<TreeNode> generateTrees(int n) {
        return generateTrees(1, n);
    }

    private List<TreeNode> generateTrees(int left, int right) {
        List<TreeNode> roots = new ArrayList<>();
        if (left > right) {
            roots.add(null);
            return roots;
        }
        for (int i = left; i <= right; i++) {
            for (TreeNode leftChild : generateTrees(left, i - 1)) {
                for (TreeNode rightChild : generateTrees(i + 1, right)) {
                    TreeNode root = new TreeNode(i);
                    root.left = leftChild;
                    root.right = rightChild;
                    roots.add(root);
                }
            }
        }
        return roots;
    }
}
